File: go/verification_test.go
package ledgerrag

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"os"
	"testing"
)

// buildSignedResponse produces a fully valid, self-signed QueryResponse over the
// given chunk texts -- exercising the whole build -> prove -> sign -> verify path.
func buildSignedResponse(t *testing.T, texts []string, prev []byte) QueryResponse {
	t.Helper()
	leaves := make([][]byte, len(texts))
	for i, s := range texts {
		leaves[i] = []byte(s)
	}
	tree, err := NewMerkleTree(leaves)
	if err != nil {
		t.Fatal(err)
	}
	priv, pub, err := GenerateKeypair()
	if err != nil {
		t.Fatal(err)
	}
	sig := SignRoot(priv, tree.Root(), prev, uint64(tree.LeafCount()))

	var prevPtr *string
	if len(prev) > 0 {
		s := hex.EncodeToString(prev)
		prevPtr = &s
	}

	citations := make([]Citation, 0, len(texts))
	proofs := make([]Proof, 0, len(texts))
	for i, s := range texts {
		cid := fmt.Sprintf("c%d", i)
		citations = append(citations, Citation{ChunkID: cid, DocID: "doc", Seq: i, Text: s})
		steps, err := tree.InclusionProof(i)
		if err != nil {
			t.Fatal(err)
		}
		wire := make([]ProofStepModel, len(steps))
		for j, ps := range steps {
			wire[j] = ProofStepModel{Sibling: hex.EncodeToString(ps.Sibling), Side: ps.Side}
		}
		sum := sha256.Sum256([]byte(s))
		proofs = append(proofs, Proof{ChunkID: cid, LeafIndex: i, SHA256: hex.EncodeToString(sum[:]), MerklePath: wire})
	}
	return QueryResponse{
		Answer:    "answer",
		Citations: citations,
		Proofs:    proofs,
		SignedRoot: SignedRoot{
			Root:      hex.EncodeToString(tree.Root()),
			PrevRoot:  prevPtr,
			LeafCount: uint64(tree.LeafCount()),
			Signature: hex.EncodeToString(sig),
			PublicKey: hex.EncodeToString(pub),
		},
	}
}

func TestSampleResponseVerifies(t *testing.T) {
	data, err := os.ReadFile("../sample-response.json")
	if err != nil {
		t.Skipf("sample-response.json unavailable: %v", err)
	}
	resp, err := ParseQueryResponse(data)
	if err != nil {
		t.Fatal(err)
	}
	result, err := VerifyResponse(resp)
	if err != nil {
		t.Fatal(err)
	}
	if !result.Verified || !result.SignatureOK {
		t.Fatalf("golden sample must verify, got %+v", result)
	}
	if len(result.Chunks) != 4 {
		t.Fatalf("expected 4 chunk results, got %d", len(result.Chunks))
	}
	for _, c := range result.Chunks {
		if !c.HashOK || !c.InclusionOK {
			t.Fatalf("chunk %s must be fully valid: %+v", c.ChunkID, c)
		}
	}
}

func TestSelfProducedResponseVerifies(t *testing.T) {
	resp := buildSignedResponse(t, []string{"alpha", "bravo", "charlie", "delta", "echo"}, nil)
	result, err := VerifyResponse(resp)
	if err != nil {
		t.Fatal(err)
	}
	if !result.Verified {
		t.Fatalf("self-produced response must verify: %+v", result)
	}
}

func TestSelfProducedResponseWithPrevRootVerifies(t *testing.T) {
	prev := make([]byte, 32)
	for i := range prev {
		prev[i] = byte(i)
	}
	resp := buildSignedResponse(t, []string{"one", "two", "three"}, prev)
	result, err := VerifyResponse(resp)
	if err != nil {
		t.Fatal(err)
	}
	if !result.Verified || !result.SignatureOK {
		t.Fatalf("response with prev_root must verify: %+v", result)
	}
}

func TestTamperedCitationTextFails(t *testing.T) {
	resp := buildSignedResponse(t, []string{"alpha", "bravo", "charlie"}, nil)
	resp.Citations[0].Text += " ...and refunds are DENIED."
	result, err := VerifyResponse(resp)
	if err != nil {
		t.Fatal(err)
	}
	if result.Verified {
		t.Fatal("tampered citation text must fail verification")
	}
	if result.Chunks[0].HashOK || result.Chunks[0].InclusionOK {
		t.Fatalf("tampered chunk should fail both hash and inclusion: %+v", result.Chunks[0])
	}
}

func TestForgedSignatureViaLeafCountFails(t *testing.T) {
	resp := buildSignedResponse(t, []string{"alpha", "bravo"}, nil)
	resp.SignedRoot.LeafCount++
	result, err := VerifyResponse(resp)
	if err != nil {
		t.Fatal(err)
	}
	if result.SignatureOK {
		t.Fatal("bumping leaf_count must break the signature")
	}
	if result.Verified {
		t.Fatal("verified must be false when the signature is broken")
	}
}

func TestHashOkUsesPlainSha256NotLeafHash(t *testing.T) {
	// Subtlety: proof.sha256 is the PLAIN SHA-256 of the chunk text, not the
	// 0x00-prefixed Merkle leaf hash. Substituting the leaf hash must fail hashOK.
	resp := buildSignedResponse(t, []string{"alpha", "bravo"}, nil)
	leafHashHex := hex.EncodeToString(HashLeaf([]byte("alpha")))
	resp.Proofs[0].SHA256 = leafHashHex
	result, err := VerifyResponse(resp)
	if err != nil {
		t.Fatal(err)
	}
	if result.Chunks[0].HashOK {
		t.Fatal("leaf hash must not satisfy the plain-sha256 integrity check")
	}
	if !result.Chunks[0].InclusionOK {
		t.Fatal("inclusion should still hold with the real Merkle path")
	}
	if result.Verified {
		t.Fatal("verified must be false when a hash check fails")
	}
}

func TestMissingCitationErrors(t *testing.T) {
	resp := buildSignedResponse(t, []string{"alpha", "bravo"}, nil)
	resp.Citations = resp.Citations[:1] // drop the citation for c1
	if _, err := VerifyResponse(resp); err == nil {
		t.Fatal("a proof without a matching citation must error")
	}
}

File: ts/src/verification.test.ts
import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";

import { describe, expect, it } from "vitest";

import { MerkleTree } from "./merkle.ts";
import type { Citation, Proof, QueryResponse } from "./models.ts";
import { generateKeypair, signRoot } from "./signing.ts";
import { verifyResponse } from "./verification.ts";

const SAMPLE_PATH = fileURLToPath(new URL("../../sample-response.json", import.meta.url));

function loadSample(): QueryResponse {
  return JSON.parse(readFileSync(SAMPLE_PATH, "utf8")) as QueryResponse;
}

/** Build a fully-signed, self-consistent response over `texts`. */
function buildSignedResponse(texts: string[], prev: Buffer | null): QueryResponse {
  const leaves = texts.map((t) => Buffer.from(t, "utf8"));
  const tree = new MerkleTree(leaves);
  const { privateKey, publicKey } = generateKeypair();
  const root = tree.root();
  const leafCount = tree.leafCount();
  const signature = signRoot(privateKey, root, prev, leafCount);

  const citations: Citation[] = texts.map((t, i) => ({
    chunk_id: `c${i}`,
    doc_id: "doc",
    seq: i,
    text: t,
  }));
  const proofs: Proof[] = texts.map((t, i) => ({
    chunk_id: `c${i}`,
    leaf_index: i,
    sha256: createHash("sha256").update(Buffer.from(t, "utf8")).digest("hex"),
    merkle_path: tree.inclusionProof(i).map((s) => ({
      sibling: s.sibling.toString("hex"),
      side: s.side,
    })),
  }));

  return {
    answer: "answer",
    citations,
    proofs,
    signed_root: {
      root: root.toString("hex"),
      prev_root: prev ? prev.toString("hex") : null,
      leaf_count: leafCount,
      signature: signature.toString("hex"),
      public_key: publicKey.toString("hex"),
    },
  };
}

describe("verification", () => {
  it("verifies the Python-produced golden sample", () => {
    const result = verifyResponse(loadSample());
    expect(result.verified).toBe(true);
    expect(result.signature_ok).toBe(true);
    expect(result.chunks).toHaveLength(4);
    for (const chunk of result.chunks) {
      expect(chunk.hash_ok).toBe(true);
      expect(chunk.inclusion_ok).toBe(true);
    }
  });

  it("verifies a self-produced response", () => {
    const resp = buildSignedResponse(["alpha", "beta", "gamma", "delta", "epsilon"], null);
    const result = verifyResponse(resp);
    expect(result.verified).toBe(true);
    expect(result.signature_ok).toBe(true);
    expect(result.chunks).toHaveLength(5);
  });

  it("verifies a self-produced response carrying a prev_root", () => {
    const resp = buildSignedResponse(["one", "two", "three"], Buffer.alloc(32, 7));
    const result = verifyResponse(resp);
    expect(result.verified).toBe(true);
    expect(result.signature_ok).toBe(true);
  });

  it("fails when a citation's text is tampered", () => {
    const resp = buildSignedResponse(["alpha", "beta", "gamma"], null);
    resp.citations[1].text = "beta-TAMPERED";
    const result = verifyResponse(resp);
    expect(result.verified).toBe(false);
    expect(result.signature_ok).toBe(true);
    const beta = result.chunks.find((c) => c.chunk_id === "c1");
    expect(beta?.hash_ok).toBe(false);
    expect(beta?.inclusion_ok).toBe(false);
  });

  it("fails when the leaf count is forged", () => {
    const resp = buildSignedResponse(["a", "b"], null);
    resp.signed_root.leaf_count += 1;
    const result = verifyResponse(resp);
    expect(result.verified).toBe(false);
    expect(result.signature_ok).toBe(false);
  });

  it("hash_ok uses the plain SHA-256 of the text, not the leaf hash", () => {
    const resp = buildSignedResponse(["alpha"], null);
    const prefixed = createHash("sha256")
      .update(Buffer.from([0x00]))
      .update(Buffer.from("alpha", "utf8"))
      .digest("hex");
    resp.proofs[0].sha256 = prefixed;
    const result = verifyResponse(resp);
    expect(result.chunks[0].hash_ok).toBe(false);
    expect(result.chunks[0].inclusion_ok).toBe(true);
  });

  it("throws when a proof references a missing citation", () => {
    const resp = buildSignedResponse(["alpha", "beta"], null);
    resp.citations.shift();
    expect(() => verifyResponse(resp)).toThrow();
  });
});
